RTC: update documentation (#64)
jexia/docs: docs/realtime/Protocol.md

# RTC protocol

The WebSocket API provides means to receive events on change resources (that the JWT has access to).

## Connecting to the WebSocket API

Build the URL according to the template below:

```
wss://<project-id>.<zone>.app.jexia.com/rtc?access_token=<jwt>
```
ARGUMENT     | TYPE   | DESCRIPTION                           | EXAMPLE
-------------|--------|---------------------------------------|----------------------------------------
project-id   | UUID   | unique identifier/hash of the project | `4f8c6ee3-f2d1-47b8-83c4-c5250b86cab4`
zone         | string | project zone                          | `nl00`
jwt          | string | valid token obtained after logging into the project using the API key credentials or any other method | `eyJhbGciOiJS... ...aev96IlwB1OFzDmyw`

> Connecting to the WebSocket API is done using a WebSocket client and connect it to the WebSocket address. Use one of the available WebSocket clients to create a connection:

Reviewer: > Use one of the available WebSocket clients to create a connection:

These are test client, so maybe you should add something like: 'you can test the WebSocket connection using one of these test clients:'?

- [Simple WebSocket Client](https://chrome.google.com/webstore/detail/simple-websocket-client/pfdhoblngboilpfeibdedpjgfnlcodoo) (Chrome extension)
- [Smart Websocket Client](https://chrome.google.com/webstore/detail/smart-websocket-client/omalebghpgejjiaoknljcfmglgbpocdp) (Chrome extension)
- [WebSocket King](https://websocketking.com) (Online WebSocket client)

> ## Before you start

Reviewer: I would put the 'Before you start' as the first section of the documentation. (it is a little weird to have a 'preparation' in the middle of the other steps)

1. Create a User/API key.

> 2. Log In providing valid credentials and method (check [here](../auth/README.md) for documentaion).

Reviewer: ```suggestion
2. Log In providing valid credentials and method (check [here](../auth/README.md) for documentation).
```

> 2. Log In providing valid credentials and method (check [here](../auth/README.md) for documentaion).

Reviewer: > ../auth/README.md

Is this the correct URL to link to another (web)page? (After the Markdown is rendered into HTML)

Author: do you mean we cannot use a relative path?
(relative path is valid if you click on the link it follows to https://github.com/jexia/docs/blob/issue/JEXIA-3985-rtc-docs/docs/auth/README.md)

3. Create some resources with WebApp or management API.

> 4. Make sure that User/API key you are going to use has all reuired permissions (policies) on the resources, create policies if not.

Reviewer: ```suggestion
4. Make sure that User/API key you are going to use has all required permissions (policies) on the resources, create policies if not.
```

## Sending and Receiving messages
After establishing the connection to the WebSocket API, the client can send messages to perform actions, like subscribing to particular events and receive messages, like notifications about changes. Both send and received messages must be using JSON encoding.

```json
{
  "type": "<message type>",
  "data": "<message data>"
}
```

ARGUMENT | TYPE   | DESCRIPTION            | EXAMPLE
---------|--------|------------------------|-----------------------------------------------
type     | String | message type           | `subscribe` / `unsubscribe` / `publish` ...
data     | Any    | message data (payload) | `{"command":"unsubscribe","arguments":{...}}`

### Command messages

```json
{
  "type": "command",
  "data": {
    "command": "<command>",
    "correlation_id": "<correlation ID: any valid JSON value>",
    "arguments": { "<command arguments>" }
  }
}
```

Each command has the `command` field in the `data` object to specify the command and optionally an `arguments` object to specify the arguments of the command.

> Field `correlation_id` is OPTIONAL and ignored by RTC, it is generated/provided by the client and returned back as apart of command response AS IS (in case it was provided by WebSocket client).

Reviewer: ```suggestion
Field `correlation_id` is OPTIONAL and ignored by RTC, it is generated/provided by the client and returned back as a part of the command response AS IS (in case it was provided by WebSocket client).
```

* `<request fields>` - object containing the command response data (if available)
* `error` - when the command failed, this object contains the error information
  * `<error code>` - Easy to use code for application to determine what went wrong (see [list of errors](#error-codes))
  * `<human-readable information>` - Textual (English) information about the error

#### Event messages
```json
{
  "type": "event",
  "data": {
    "action": "<action>",
    "resource": {
      "type": "<resource type>",
      "name": "<resource name>",
    },
    "modifier": {
      "id": "<modifier identifier>",
    },
    "timestamp": "<timestamp of change>",
    "data": [{ "<event data>" }]
  }
}
```

* `<action>`, `<resource type>`, `<resource name>` - Describe the event that happened on which resource (see [Events](#events) for more information).
* `<modifier identifier>` - Optional field describing who/what changed the event (see [Events](#events) for more information).
* `<timestamp>` - ISO 8601 timestamp as an indication when the event occurred.
* `<event data>` contains the data of the event, which includes only the `ID` field for the moment.

#### Notification messages

> 

Reviewer: Maybe add a (single) line to explain what notification are?

### Unsubscribe from resource event

Send the following command to unsubscribe to the specified event(s).
```json
{
  "type": "command",
  "data": {
    "command": "unsubscribe",
    "correlation_id": 43,
    "arguments": {
      "action": [ "<action(s)>" ],
      "resource": {
        "type": "<resource type>",
        "name": "<resource name>"
      }
    }
  }
}
```

The response indicates if subscribing succeeded or not, typical [error responses](#error-codes) are 1000 (Unknown resource type), 1001 (Unknown resource name), and 1002 (Unknown action).

## Channels

RTC module supports event channels that allow applications to publish custom events/payloads.
Technically they are similar to events the only difference you have to subscribe to the `published` action:

```json
{
  "type": "command",
  "data": {
    "command": "subscribe",
    "correlation_id": "some unique value",
    "arguments": {
      "action": [ "published" ],
      "resource": {
        "type": "channel",
        "name": "example"
      }
    }
  }
}
```

### Publishing messages

> 1. Use WebApp to create a channel or check the [API documentation](../mgm/README.md) to learn more about management operations on channels.

Reviewer: check URL
